2.0-proposal/RELATIONS.md

```diff
@@ -0,0 +1,248 @@
+```
+
+In the relatively common case that you don't expect or care about multiple
+related applications, you can use the `self.all_units` view, which is roughly
+analogous to `[u for a in self.applications for u in a.units]` except that it
+also supports the filter and flag setting / removing methods.
+
+It's important to note that each application in the collection actually
+represents a relation, and so a given remote application might actually appear
+in the collection multiple times if it is related more than once (for example,
+if an application needed multiple databases).  However, the collection of units
+and flags associated with the application or units will be kept separate.  You
+can distinguish between the application instances by their `relation_id`
+property.
+
+## Relation Data
+
+As shown above, each unit has a data attribute.  This behaves like a
+read-only `defaultdict(None)` which allows the keys to be accessed as
+attributes (with dashes converted to underscores).  The values are always
+strings or `None`.
+
+There is also a helper view for the relation data set by remote units, in case
+you expect only the leader to be providing data or for the remote units to
+agree on (some subset) of the data.  For example:
+
+```python
+class MySQLClient(SimpleRelation):
+    class flags:
+        changed = 'relation.{relation_name}.changed'
+        ready = 'relation.{relation_name}.ready'
+
+    @when(flags.changed)
+    def changed(self):
+        data = self.all_units.merged_data
+        # we assume data.host to be the master
+        if all([data.host, data.database, data.user, data.password]):
+            self.add_flag(self.flags.ready)
+```
+
+To set relation data for the current unit, there is a `set_data` collection on
+each application which behaves like the read-only data collection, but which
+allows setting attributes or dictionary values.  For example:
+
+```python
+class MySQL(SimpleRelation):
+    class flags:
+        provided = 'relation.{relation_name}.provided'
+
+    def requested_databases(self):
+        return self.applications.without_flag(self.flags.provided)
+
+    def provide_database(self, app, host, port, db, user, pass):
+        app.set_data.host = host
+        app.set_data.port = port
+        app.set_data.database = db
+        app.set_data.user = user
+        app.set_data.password = pass
+        app.add_flag(self.flags.provided)
+```
+
+It is often useful to have typed or structured data, which involves encoding
+and decoding to something like JSON.  To make this easy, the data collections
+support a JSON version which do automatic JSON encoding / decoding.
+
+```python
+class MyServicePeer(SimpleRelation):
+    class flags:
+        joined = 'relation.{relation_name}.joined'
+        changed = 'relation.{relation_name}.changed'
+
+    @when(flags.joined)
+    def send_data(self):
+        for app in self.applications:
+            app.set_json_data.list = ['one', 'two']
```

Comment:
@johnsca Small note: I suspect `list` is a key and `['one', 'two']` is the value? Then it's best to call them smth like "my_list" and "my_bool" to make sure that `list` is a variable name and not a type.